tensorflow-framework/src/main/java/org/tensorflow/framework/losses/impl/LossesImpl.java

```diff
@@ -0,0 +1,304 @@
+package org.tensorflow.framework.losses.impl;
+
+import org.tensorflow.DataType;
+import org.tensorflow.Operand;
+import org.tensorflow.framework.losses.Reduction;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.op.Ops;
+import org.tensorflow.op.core.ReduceSum;
+import org.tensorflow.op.core.Squeeze;
+import org.tensorflow.types.TInt32;
+import org.tensorflow.types.family.TNumber;
+
+import java.util.Collections;
+
+public class LossesImpl {
+
+  /**
+   * Squeeze or expand last dimension if needed with a sampleWeights of one.
+   *
+   * <ol type="1">
+   *   <li>Squeezes last dim of <code>predictions</code> or <code>labels</code> if their rank differs by 1 (using
+   *       {@link #removeSqueezableDimensions}).
+   *   <li>Squeezes or expands last dim of <code>sampleWeight` if its rank differs by 1 from the new
+   *       rank of <code>predictions`. If <code>sampleWeight` is scalar, it is kept scalar./li>
+   * </ol>
+   *
+   * @param tf the TensorFlow Ops
+   * @param predictions Predicted values, a <code>Operand</code> of arbitrary dimensions.
+   * @param labels Optional label <code>Operand</code> whose dimensions match <code>prediction</code>.
+   * @return Tuple of <code>prediction</code>, <code>label</code> and <code>sampleWeight</code>. Each of them possibly has the last
+   *     dimension squeezed, <code>sampleWeight</code> could be extended by one dimension. If <code>sampleWeight</code>
+   *     is null, (prediction, label) is returned.
+   */
+  public static <T extends TNumber> Tuple<T> squeezeOrExpandDimensions(
+      Ops tf, Operand<T> labels, Operand<T> predictions) {
+    return squeezeOrExpandDimensions(tf, labels, predictions, null);
+  }
+
+  /**
+   * Squeeze or expand last dimension if needed.
+   *
+   * <ol type="1">
+   *   <li>Squeezes last dim of `predictions` or `labels` if their rank differs by 1 (using *
+   *       `confusion_matrix.remove_squeezable_dimensions`). *
+   *   <li>Squeezes or expands last dim of `sampleWeight` if its rank differs by 1 from the new *
+   *       rank of `predictions`. If `sampleWeight` is scalar, it is kept scalar./li> *
+   * </ol>
+   *
+   * @param tf the TensorFlow Ops
+   * @param predictions Predicted values, a <code>Operand</code> of arbitrary dimensions.
+   * @param labels Optional label <code>Operand</code> whose dimensions match <code>prediction
+   *     </code>.
+   * @param sampleWeight Optional sample weight(s) <code>Operand</code> whose dimensions match<code>
+   *     prediction</code>.
+   * @return Tuple of <code>prediction</code>, <code>label</code> and <code>sampleWeight</code>.
+   *     Each of them possibly has the last dimension squeezed, <code>sampleWeight</code> could be
+   *     extended by one dimension. If <code>sampleWeight</code> is null, (prediction, label) is
+   *     returned.
+   */
```

Comment:
This method has a myriad of complex cases, so I think it deserves its own direct unit test.

Reply:
I could not find direct test cases for this method in Python. It's defined in `tensorflow/tensorflow/python/ops/losses/utils.py`. You want to take a stab at it?

Reply:
:-) Totally. I want to do some work on #92 first, so I'll open an issue for myself.